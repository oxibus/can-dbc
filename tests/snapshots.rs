#![cfg(all(feature = "serde", feature = "encodings"))]

use nom::Err;
use std::borrow::Cow;
use std::fs;
use std::path::{Path, PathBuf};

use can_dbc::{decode_cp1252, Dbc, Error};
use insta::{assert_debug_snapshot, assert_yaml_snapshot, with_settings};
use test_each_file::test_each_path;

struct TestConfig {
    test_root: &'static str,
    snapshot_suffix: &'static str,
    use_cp1251: bool,
    create_snapshot: bool,
}

static TEST_DIRS: &[TestConfig] = &[
    TestConfig {
        test_root: "shared-test-files",
        snapshot_suffix: "",
        use_cp1251: true,
        create_snapshot: true,
    },
    TestConfig {
        test_root: "opendbc/opendbc/dbc",
        snapshot_suffix: "opendbc",
        use_cp1251: false,
        create_snapshot: false,
    },
];

test_each_path! { for ["dbc"] in "./tests/fixtures/opendbc/opendbc/dbc" as dbc => parse_one_file }
test_each_path! { for ["dbc"] in "./tests/fixtures/shared-test-files" as shared => parse_one_file }
// upper case extension
test_each_path! { for ["DBC"] in "./tests/fixtures/shared-test-files" as shared2 => parse_one_file }

/// Get `test root`, `snapshot name suffix`, `use cp1251`, `create snapshot` for the given path
fn get_test_info(path: &Path) -> Option<(PathBuf, &'static TestConfig)> {
    if !path
        .extension()
        .unwrap_or_default()
        .eq_ignore_ascii_case("dbc")
    {
        return None;
    }
    let path_str = path.to_str().unwrap();
    let parent = path.parent().unwrap();
    for item in TEST_DIRS {
        // Ensure slashes are there for easier matching
        let test_root = format!("/{}/", item.test_root);
        let mut path_dir = parent.to_str().unwrap().to_string();
        if !path_dir.ends_with('/') {
            path_dir.push('/');
        }
        if let Some(pos) = path_dir.find(&test_root) {
            let parent = PathBuf::from("snapshots")
                .join(item.snapshot_suffix)
                .join(&path_dir[pos + test_root.len()..]);
            return Some((parent, item));
        }
    }
    panic!("Unknown test directory: {path_str}");
}

/// Test parsing all DBC files
#[test]
fn test_if_submodules_are_present() {
    for test in TEST_DIRS {
        let dir = Path::new("./tests/fixtures").join(test.test_root);
        fs::read_dir(&dir).unwrap_or_else(|e| {
            let dir_display = dir.display();
            panic!(
                "
--------------------------------------------------------------------------
Error reading dbc test files from   {dir_display}
{e}
Make sure git submodules are up to date by running
    git submodule update --init --recursive
--------------------------------------------------------------------------
"
            )
        });
    }
}

/// Parse a single DBC file and assert a snapshot of the result.
fn parse_one_file([path]: [&Path; 1]) {
    let Some((
        snapshot_path,
        &TestConfig {
            use_cp1251,
            create_snapshot,
            ..
        },
    )) = get_test_info(path)
    else {
        return;
    };

    let file_name = path.file_stem().unwrap().to_string_lossy().to_string();
    let buffer = fs::read(path).unwrap();
    let buffer = if use_cp1251 {
        decode_cp1252(&buffer)
            .unwrap_or_else(|| panic!("Failed to decode {} as cp1252", path.display()))
    } else {
        Cow::Borrowed(
            std::str::from_utf8(&buffer)
                .unwrap_or_else(|_| panic!("Failed to decode {} as utf-8", path.display())),
        )
    };

    report_test_result(
        Dbc::try_from(buffer.as_ref()),
        snapshot_path,
        file_name,
        create_snapshot,
    );
}

fn report_test_result(
    result: Result<Dbc, Error>,
    snapshot_path: PathBuf,
    file_name: String,
    do_snapshot: bool,
) {
    let do_snapshot = std::env::var("SKIP_INSTA").is_err()
        && (do_snapshot || std::env::var("FORCE_INSTA").is_ok());

    with_settings! {
        {
            omit_expression => true,
            snapshot_path => snapshot_path,
            prepend_module_to_snapshot => false
        },
        {
            match result {
                Ok(result) => {
                    if do_snapshot {
                        assert_yaml_snapshot!(file_name, result);
                    } // else ignore successful results when not snapshotting
                }
                Err(e) => {
                    let file_name = format!("!error___{file_name}");
                    assert_debug_snapshot!(file_name, err_to_str(&e));
                }
            }
        }
    }
}

fn err_to_str<'a>(err: &'a Error) -> &'a str {
    match err {
        Error::Incomplete(_, _) => "incomplete",
        Error::Nom(e) => match e {
            Err::Incomplete(_) => "nom-incomplete",
            Err::Error(_) => "nom-error",
            Err::Failure(_) => "nom-failure",
        },
        Error::MultipleMultiplexors => "multiple multiplexors",
    }
}
