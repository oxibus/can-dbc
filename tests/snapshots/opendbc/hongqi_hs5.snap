---
source: tests/snapshots.rs
---
version: ""
new_symbols:
  - NS_DESC_
  - CM_
  - BA_DEF_
  - BA_
  - VAL_
  - CAT_DEF_
  - CAT_
  - FILTER
  - BA_DEF_DEF_
  - EV_DATA_
  - ENVVAR_DATA_
  - SGTYPE_
  - SGTYPE_VAL_
  - BA_DEF_SGTYPE_
  - BA_SGTYPE_
  - SIG_TYPE_REF_
  - VAL_TABLE_
  - SIG_GROUP_
  - SIG_VALTYPE_
  - SIGTYPE_VALTYPE_
  - BO_TX_BU_
  - BA_DEF_REL_
  - BA_REL_
  - BA_DEF_DEF_REL_
  - BU_SG_REL_
  - BU_EV_REL_
  - BU_BO_REL_
  - SG_MUL_VAL_
bit_timing: []
nodes:
  - XXX
messages:
  - id:
      Standard: 146
    name: ECM_1
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: CHECKSUM
        multiplexer_indicator: Plain
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: THROTTLE_1
        multiplexer_indicator: Plain
        start_bit: 8
        size: 12
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: THROTTLE_2
        multiplexer_indicator: Plain
        start_bit: 24
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: THROTTLE_3
        multiplexer_indicator: Plain
        start_bit: 32
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: DRIVER_THROTTLE
        multiplexer_indicator: Plain
        start_bit: 40
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: COUNTER
        multiplexer_indicator: Plain
        start_bit: 60
        size: 4
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 192
    name: ABS_1
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: CHECKSUM
        multiplexer_indicator: Plain
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
      - name: FRONT_LEFT
        multiplexer_indicator: Plain
        start_bit: 8
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.01
        offset: 0
        min: 0
        max: 65535
        unit: km/h
        receivers:
          - XXX
      - name: FRONT_RIGHT
        multiplexer_indicator: Plain
        start_bit: 24
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.01
        offset: 0
        min: 0
        max: 65535
        unit: km/h
        receivers:
          - XXX
      - name: VEHICLE_SPEED
        multiplexer_indicator: Plain
        start_bit: 40
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.01
        offset: 0
        min: 0
        max: 65535
        unit: km/h
        receivers:
          - XXX
      - name: COUNTER
        multiplexer_indicator: Plain
        start_bit: 60
        size: 4
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 194
    name: ABS_2
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: CHECKSUM
        multiplexer_indicator: Plain
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
      - name: REAR_LEFT
        multiplexer_indicator: Plain
        start_bit: 8
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.01
        offset: 0
        min: 0
        max: 65535
        unit: km/h
        receivers:
          - XXX
      - name: REAR_RIGHT
        multiplexer_indicator: Plain
        start_bit: 24
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.01
        offset: 0
        min: 0
        max: 65535
        unit: km/h
        receivers:
          - XXX
      - name: BRAKE_PRESSURE
        multiplexer_indicator: Plain
        start_bit: 40
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: COUNTER
        multiplexer_indicator: Plain
        start_bit: 60
        size: 4
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 208
    name: EPS_1
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: CHECKSUM
        multiplexer_indicator: Plain
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
      - name: STEER_ANGLE
        multiplexer_indicator: Plain
        start_bit: 8
        size: 15
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.0225
        offset: 0
        min: 0
        max: 65535
        unit: ""
        receivers:
          - XXX
      - name: STEER_ANGLE_DIRECTION
        multiplexer_indicator: Plain
        start_bit: 23
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: direction
        receivers:
          - XXX
      - name: STEER_RATE
        multiplexer_indicator: Plain
        start_bit: 24
        size: 15
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 0.0225
        offset: 0
        min: 0
        max: 65535
        unit: ""
        receivers:
          - XXX
      - name: STEER_RATE_DIRECTION
        multiplexer_indicator: Plain
        start_bit: 39
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: direction
        receivers:
          - XXX
      - name: COUNTER
        multiplexer_indicator: Plain
        start_bit: 60
        size: 4
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 336
    name: EPS_2
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: CHECKSUM
        multiplexer_indicator: Plain
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: EPS_TORQUE
        multiplexer_indicator: Plain
        start_bit: 8
        size: 10
        byte_order: LittleEndian
        value_type: Signed
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: EPS_TORQUE_DIRECTION
        multiplexer_indicator: Plain
        start_bit: 18
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: direction
        receivers:
          - XXX
      - name: UNKNOWN
        multiplexer_indicator: Plain
        start_bit: 22
        size: 2
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 3
        unit: ""
        receivers:
          - XXX
      - name: LKAS_STATUS
        multiplexer_indicator: Plain
        start_bit: 24
        size: 4
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
      - name: DRIVER_INPUT_TORQUE
        multiplexer_indicator: Plain
        start_bit: 32
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: LKAS_TORQUE
        multiplexer_indicator: Plain
        start_bit: 40
        size: 10
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1023
        unit: ""
        receivers:
          - XXX
      - name: LKAS_TORQUE_DIRECTION
        multiplexer_indicator: Plain
        start_bit: 50
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
      - name: COUNTER
        multiplexer_indicator: Plain
        start_bit: 60
        size: 4
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 272
    name: ACC
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: CHECKSUM
        multiplexer_indicator: Plain
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: UNKNOWN_1
        multiplexer_indicator: Plain
        start_bit: 8
        size: 12
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 4095
        unit: ""
        receivers:
          - XXX
      - name: UNKNOWN_2
        multiplexer_indicator: Plain
        start_bit: 20
        size: 4
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
      - name: ACCELERATION
        multiplexer_indicator: Plain
        start_bit: 24
        size: 12
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: -300
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: STATUS
        multiplexer_indicator: Plain
        start_bit: 36
        size: 4
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 31
        unit: x
        receivers:
          - XXX
      - name: UNKNOWN_3
        multiplexer_indicator: Plain
        start_bit: 46
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
      - name: COUNTER
        multiplexer_indicator: Plain
        start_bit: 60
        size: 4
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 274
    name: LKAS
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: CHECKSUM
        multiplexer_indicator: Plain
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: LKAS_TORQUE
        multiplexer_indicator: Plain
        start_bit: 8
        size: 10
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 2047
        unit: x
        receivers:
          - XXX
      - name: LKAS_TORQUE_DIRECTION
        multiplexer_indicator: Plain
        start_bit: 18
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
      - name: LKAS_TORQUE_ACTIVE
        multiplexer_indicator: Plain
        start_bit: 19
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
      - name: UNKNOWN_ACTIVE_STATE
        multiplexer_indicator: Plain
        start_bit: 26
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
      - name: MAYBE_HUD_LANE_STATES
        multiplexer_indicator: Plain
        start_bit: 28
        size: 2
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 3
        unit: ""
        receivers:
          - XXX
      - name: LKAS_ACTIVE
        multiplexer_indicator: Plain
        start_bit: 32
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: COUNTER
        multiplexer_indicator: Plain
        start_bit: 60
        size: 4
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 148
    name: MAYBE_ABS
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: CHECKSUM
        multiplexer_indicator: Plain
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: UNKNOWN_1
        multiplexer_indicator: Plain
        start_bit: 8
        size: 16
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: x
        receivers:
          - XXX
      - name: UNKNOWN_2
        multiplexer_indicator: Plain
        start_bit: 24
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: BRAKE_PRESSED
        multiplexer_indicator: Plain
        start_bit: 35
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: x
        receivers:
          - XXX
      - name: UNKNOWN_3
        multiplexer_indicator: Plain
        start_bit: 46
        size: 2
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 3
        unit: ""
        receivers:
          - XXX
      - name: COUNTER
        multiplexer_indicator: Plain
        start_bit: 60
        size: 4
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 304
    name: ABS_3
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: CHECKSUM
        multiplexer_indicator: Plain
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
      - name: NEW_SIGNAL_1
        multiplexer_indicator: Plain
        start_bit: 15
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: NEW_SIGNAL_2
        multiplexer_indicator: Plain
        start_bit: 23
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: NEW_SIGNAL_3
        multiplexer_indicator: Plain
        start_bit: 31
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: NEW_SIGNAL_4
        multiplexer_indicator: Plain
        start_bit: 39
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: NEW_SIGNAL_5
        multiplexer_indicator: Plain
        start_bit: 47
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: DRIVER_BRAKE_PRESSURE
        multiplexer_indicator: Plain
        start_bit: 48
        size: 12
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 65535
        unit: ""
        receivers:
          - XXX
      - name: COUNTER
        multiplexer_indicator: Plain
        start_bit: 60
        size: 4
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 144
    name: MAYBE_ENGINE
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: CHECKSUM
        multiplexer_indicator: Plain
        start_bit: 0
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: NEW_SIGNAL_4
        multiplexer_indicator: Plain
        start_bit: 8
        size: 12
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: MAYBE_RPM
        multiplexer_indicator: Plain
        start_bit: 20
        size: 12
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
      - name: NEW_SIGNAL_1
        multiplexer_indicator: Plain
        start_bit: 32
        size: 8
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 255
        unit: ""
        receivers:
          - XXX
      - name: NEW_SIGNAL_2
        multiplexer_indicator: Plain
        start_bit: 46
        size: 14
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 4095
        unit: ""
        receivers:
          - XXX
      - name: COUNTER
        multiplexer_indicator: Plain
        start_bit: 60
        size: 4
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 15
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 422
    name: MAYBE_ACC_BUTTONS
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: NEW_SIGNAL_1
        multiplexer_indicator: Plain
        start_bit: 18
        size: 1
        byte_order: BigEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 560
    name: TURN_SIGNALS
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: RIGHT_TURN_SIGNALING
        multiplexer_indicator: Plain
        start_bit: 45
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
      - name: LEFT_TURN_SIGNALING
        multiplexer_indicator: Plain
        start_bit: 44
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
      - name: RIGHT_TURN_BULB_3
        multiplexer_indicator: Plain
        start_bit: 21
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
      - name: LEFT_TURN_BULB_3
        multiplexer_indicator: Plain
        start_bit: 20
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
      - name: RIGHT_TURN_BULB_2
        multiplexer_indicator: Plain
        start_bit: 19
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
      - name: LEFT_TURN_BULB_2
        multiplexer_indicator: Plain
        start_bit: 18
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
      - name: LEFT_TURN_BULB_1
        multiplexer_indicator: Plain
        start_bit: 16
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
      - name: RIGHT_TURN_BULB_1
        multiplexer_indicator: Plain
        start_bit: 17
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 586
    name: DOOR_FL
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: OPEN
        multiplexer_indicator: Plain
        start_bit: 2
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 588
    name: DOOR_FR
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: OPEN
        multiplexer_indicator: Plain
        start_bit: 2
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 590
    name: DOOR_RL
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: OPEN
        multiplexer_indicator: Plain
        start_bit: 2
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
  - id:
      Standard: 591
    name: DOOR_RR
    size: 8
    transmitter:
      NodeName: XXX
    signals:
      - name: OPEN
        multiplexer_indicator: Plain
        start_bit: 2
        size: 1
        byte_order: LittleEndian
        value_type: Unsigned
        factor: 1
        offset: 0
        min: 0
        max: 1
        unit: ""
        receivers:
          - XXX
comments:
  - Signal:
      message_id:
        Standard: 274
      name: UNKNOWN_ACTIVE_STATE
      comment: Triggers changes from 8 to 9 in EPS_2.LKAS_STATUS
  - Signal:
      message_id:
        Standard: 560
      name: RIGHT_TURN_SIGNALING
      comment: Includes 2.5 second convenience lane change interval
  - Signal:
      message_id:
        Standard: 560
      name: LEFT_TURN_SIGNALING
      comment: Includes 2.5 second convenience lane change interval
value_descriptions:
  - Signal:
      message_id:
        Standard: 336
      name: LKAS_STATUS
      value_descriptions:
        - id: 1
          description: INITIALIZING
        - id: 5
          description: READY
        - id: 8
          description: ACTIVE_1
        - id: 9
          description: ACTIVE_2
        - id: 13
          description: FAULT
  - Signal:
      message_id:
        Standard: 586
      name: OPEN
      value_descriptions:
        - id: 0
          description: CLOSED
        - id: 1
          description: OPEN
  - Signal:
      message_id:
        Standard: 588
      name: OPEN
      value_descriptions:
        - id: 0
          description: CLOSED
        - id: 1
          description: OPEN
  - Signal:
      message_id:
        Standard: 590
      name: OPEN
      value_descriptions:
        - id: 0
          description: CLOSED
        - id: 1
          description: OPEN
  - Signal:
      message_id:
        Standard: 591
      name: OPEN
      value_descriptions:
        - id: 0
          description: CLOSED
        - id: 1
          description: OPEN
